GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs

```diff
@@ -1155,11 +1119,15 @@ private void UpdatePlaceholders()
                 int deleteFolderPlaceholderAttempted = 0;
                 using (BlobSizes.BlobSizesConnection blobSizesConnection = this.blobSizes.CreateConnection())
                 {
-                    // A hash of the folder placeholders is only required if the platform expands directories
-                    HashSet<string> folderPlaceholders =
-                        GVFSPlatform.Instance.KernelDriver.EnumerationExpandsDirectories ?
-                        new HashSet<string>(placeholderFoldersListCopy.Select(x => x.Path), StringComparer.OrdinalIgnoreCase) :
-                        null;
+                    // A hash of the placeholders is only required if the platform expands directories
+                    HashSet<string> existingPlaceholders = null;
+                    if (GVFSPlatform.Instance.KernelDriver.EnumerationExpandsDirectories)
+                    {
+                        IEnumerable<string> allPlaceholders = placeholderFoldersListCopy
+                            .Select(x => x.Path)
+                            .Union(this.placeholderDatabase.GetAllFilePaths());

```

Comment:
>Do we have to get all placeholders this way for correctness?

No

>Or could we get get all of the entries out of the database?

We could but we already got all the folder placeholders and they have not changed since only the file placeholders were processed up to this point and if there are are a significant number of folder placeholders it will be slower to get them all from the database again.  I will add some comments here to help explain this.